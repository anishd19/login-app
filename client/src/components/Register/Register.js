import React, { Component } from "react";
import { Redirect, Link } from "react-router-dom";
import { connect } from "react-redux";
import { registerUser } from "../../actions/authActions";
import ListErrors from "../common/ListErrors";
import "./Register.scss";

import { Logo } from "../Logo";

const countries = ["India", "US", "UK", "China"];

class Register extends Component {
  constructor() {
    super();
    this.state = {
      redirectToReferrer: false,
      username: "",
      email: "",
      password: "",
      firstname: "",
      lastname: "",
      gender: "male",
      country: countries[0],
      errors: {}
    };

    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  componentDidMount() {
    if (this.props.auth && this.props.auth.isAuthenticated) {
      this.setState({ redirectToReferrer: true });
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.auth.isAuthenticated) {
      this.setState({ redirectToReferrer: true });
    }

    if (nextProps.errors) {
      this.setState({ errors: nextProps.errors });
    }
  }

  onSubmit(e) {
    e.preventDefault();

    let {
      username,
      email,
      password,
      firstname,
      lastname,
      gender,
      country
    } = this.state;

    const userData = {
      user: {
        username,
        email,
        password,
        firstname,
        lastname,
        gender,
        country
      }
    };

    this.props.registerUser(userData, this.props.history);
  }

  onChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }
  render() {
    const { from } = this.props.location.state || { from: { pathname: "/" } };
    const { redirectToReferrer } = this.state;

    let {
      username,
      email,
      password,
      firstname,
      lastname,
      gender,
      country
    } = this.state;

    if (redirectToReferrer) {
      return <Redirect to={from} />;
    }

    return (
      <section className="hero is-fullheight landing">
        <div className="hero-body">
          <div className="container has-text-centered" id="login-form">
            <div className="box">
              <div>
                <Logo />
              </div>
              <h3 className="title">Register</h3>
              <ListErrors errors={this.props.errors} />
              <form>
                <div className="field">
                  <p className="control has-icons-left has-icons-right">
                    <input
                      className="input"
                      type="text"
                      name="username"
                      placeholder="Username"
                      value={username}
                      onChange={this.onChange}
                    />
                    <span className="icon is-small is-left">
                      <i className="fas fa-user"></i>
                    </span>
                  </p>
                </div>
                <div className="field">
                  <p className="control has-icons-left has-icons-right">
                    <input
                      className="input"
                      type="text"
                      name="firstname"
                      placeholder="firstname"
                      value={firstname}
                      onChange={this.onChange}
                    />
                    <span className="icon is-small is-left">
                      <i className="fas fa-user"></i>
                    </span>
                  </p>
                </div>
                <div className="field">
                  <p className="control has-icons-left has-icons-right">
                    <input
                      className="input"
                      type="text"
                      name="lastname"
                      placeholder="lastname"
                      value={lastname}
                      onChange={this.onChange}
                    />
                    <span className="icon is-small is-left">
                      <i className="fas fa-user"></i>
                    </span>
                  </p>
                </div>
                <div className="field">
                  <p className="control has-icons-left has-icons-right">
                    <input
                      className="input"
                      type="email"
                      name="email"
                      placeholder="Email"
                      value={email}
                      onChange={this.onChange}
                    />
                    <span className="icon is-small is-left">
                      <i className="fas fa-envelope"></i>
                    </span>
                  </p>
                </div>
                <div className="field is-horizontal is-pulled-left">
                  <div className="field-label is-normal">
                    <label className="label">Gender</label>
                  </div>
                  <div className="field-body">
                    <div className="field radio-field">
                      <div className="control">
                        <label className="radio">
                          <input
                            type="radio"
                            name="gender"
                            value="male"
                            checked={gender === "male"}
                            onChange={this.onChange}
                          />
                          Male
                        </label>
                        <label className="radio">
                          <input
                            type="radio"
                            name="gender"
                            value="female"
                            checked={gender === "female"}
                            onChange={this.onChange}
                          />
                          Female
                        </label>
                      </div>
                    </div>
                  </div>
                </div>
                <div className="field is-horizontal is-pulled-right">
                  <div className="field-label is-normal">
                    <label className="label">Country</label>
                  </div>
                  <div className="field-body">
                    <div className="field is-narrow">
                      <div className="control">
                        <div className="select is-fullwidth">
                          <select value={country} name="country" onChange={this.onChange}>
                            {countries.map(country => {
                              return <option
                                        value={country}
                                        key={country}>
                                        {country}
                                        </option>;
                            })}
                          </select>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div className="field">
                  <p className="control has-icons-left">
                    <input
                      className="input"
                      type="password"
                      name="password"
                      placeholder="Password"
                      value={password}
                      onChange={this.onChange}
                    />
                    <span className="icon is-small is-left">
                      <i className="fas fa-lock"></i>
                    </span>
                  </p>
                </div>
                <button className="button is-block is-info is-large is-fullwidth" onClick={this.onSubmit}>
                  Submit
                </button>
              </form>
              <div>
                <p className="has-text-grey sub-menu">
                  <Link to="/login">Login</Link> &nbsp;·&nbsp;
                  <a href="../">Forgot Password</a> &nbsp;·&nbsp;
                  <a href="../">Need Help?</a>
                </p>
              </div>
            </div>
          </div>
        </div>
      </section>
    );
  }
}

const mapStateToProps = state => ({
  auth: state.auth,
  errors: state.errors
});

export default connect(mapStateToProps, { registerUser })(Register);
