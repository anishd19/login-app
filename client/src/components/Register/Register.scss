#login-form {
  max-width: 500px;

  .radio-field {
    margin-top: 6px;
  }
}
