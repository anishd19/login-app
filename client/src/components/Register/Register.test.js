import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { cleanup, fireEvent, render } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import Register from "./Register";
import store from "../../store";

afterEach(cleanup);

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Router>
          <Register location={false} />
        </Router>
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders html elements", () => {
  const { getByText } = render(
    <Provider store={store}>
      <Router>
        <Register location={false} />
      </Router>
    </Provider>
  );

  expect(getByText(/Forgot Password/i)).toBeTruthy();
  expect(getByText(/Need Help?/i)).toBeTruthy();
});

// it("Enter email and password and submit", () => {
//   const { getByPlaceholderText, getByLabelText } = render(
//     <Router>
//       <Login />
//     </Router>
//   );

//   userEvent.type(getByPlaceholderText("Username"), "anishd19");
//   expect(getByPlaceholderText("Username")).toHaveAttribute("value", "anishd19");

//   userEvent.type(getByPlaceholderText("firstname"), "Anish");
//   expect(getByPlaceholderText("firstname")).toHaveAttribute("value", "Anish");

//   userEvent.type(getByPlaceholderText("lastname"), "Dhamodaran");
//   expect(getByPlaceholderText("lastname")).toHaveAttribute(
//     "value",
//     "Dhamodaran"
//   );

//   userEvent.type(getByPlaceholderText("Email"), "anishd19@gmail.com");
//   expect(getByPlaceholderText("Email")).toHaveAttribute(
//     "value",
//     "anishd19@gmail.com"
//   );

//   const radio = getByLabelText("Male");
//   fireEvent.change(radio, { target: { value: "Female" } });
//   expect(radio.value).toBe("Female");

//   userEvent.type(getByPlaceholderText("Password"), "secretPass123");
//   expect(getByPlaceholderText("Password")).toHaveAttribute(
//     "value",
//     "secretPass123"
//   );

//   // fireEvent.click(getByText(/Submit/i));
// });
