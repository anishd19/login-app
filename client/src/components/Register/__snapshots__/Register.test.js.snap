// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="hero is-fullheight landing"
>
  <div
    className="hero-body"
  >
    <div
      className="container has-text-centered"
      id="login-form"
    >
      <div
        className="box"
      >
        <div>
          <div
            className="logo"
          >
            Login App
          </div>
        </div>
        <h3
          className="title"
        >
          Register
        </h3>
        <ul
          className="error-messages help is-danger"
        />
        <form>
          <div
            className="field"
          >
            <p
              className="control has-icons-left has-icons-right"
            >
              <input
                className="input"
                name="username"
                onChange={[Function]}
                placeholder="Username"
                type="text"
                value=""
              />
              <span
                className="icon is-small is-left"
              >
                <i
                  className="fas fa-user"
                />
              </span>
            </p>
          </div>
          <div
            className="field"
          >
            <p
              className="control has-icons-left has-icons-right"
            >
              <input
                className="input"
                name="firstname"
                onChange={[Function]}
                placeholder="firstname"
                type="text"
                value=""
              />
              <span
                className="icon is-small is-left"
              >
                <i
                  className="fas fa-user"
                />
              </span>
            </p>
          </div>
          <div
            className="field"
          >
            <p
              className="control has-icons-left has-icons-right"
            >
              <input
                className="input"
                name="lastname"
                onChange={[Function]}
                placeholder="lastname"
                type="text"
                value=""
              />
              <span
                className="icon is-small is-left"
              >
                <i
                  className="fas fa-user"
                />
              </span>
            </p>
          </div>
          <div
            className="field"
          >
            <p
              className="control has-icons-left has-icons-right"
            >
              <input
                className="input"
                name="email"
                onChange={[Function]}
                placeholder="Email"
                type="email"
                value=""
              />
              <span
                className="icon is-small is-left"
              >
                <i
                  className="fas fa-envelope"
                />
              </span>
            </p>
          </div>
          <div
            className="field is-horizontal is-pulled-left"
          >
            <div
              className="field-label is-normal"
            >
              <label
                className="label"
              >
                Gender
              </label>
            </div>
            <div
              className="field-body"
            >
              <div
                className="field radio-field"
              >
                <div
                  className="control"
                >
                  <label
                    className="radio"
                  >
                    <input
                      checked={true}
                      name="gender"
                      onChange={[Function]}
                      type="radio"
                      value="male"
                    />
                    Male
                  </label>
                  <label
                    className="radio"
                  >
                    <input
                      checked={false}
                      name="gender"
                      onChange={[Function]}
                      type="radio"
                      value="female"
                    />
                    Female
                  </label>
                </div>
              </div>
            </div>
          </div>
          <div
            className="field is-horizontal is-pulled-right"
          >
            <div
              className="field-label is-normal"
            >
              <label
                className="label"
              >
                Country
              </label>
            </div>
            <div
              className="field-body"
            >
              <div
                className="field is-narrow"
              >
                <div
                  className="control"
                >
                  <div
                    className="select is-fullwidth"
                  >
                    <select
                      name="country"
                      onChange={[Function]}
                      value="India"
                    >
                      <option
                        value="India"
                      >
                        India
                      </option>
                      <option
                        value="US"
                      >
                        US
                      </option>
                      <option
                        value="UK"
                      >
                        UK
                      </option>
                      <option
                        value="China"
                      >
                        China
                      </option>
                    </select>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="field"
          >
            <p
              className="control has-icons-left"
            >
              <input
                className="input"
                name="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
              />
              <span
                className="icon is-small is-left"
              >
                <i
                  className="fas fa-lock"
                />
              </span>
            </p>
          </div>
          <button
            className="button is-block is-info is-large is-fullwidth"
            onClick={[Function]}
          >
            Submit
          </button>
        </form>
        <div>
          <p
            className="has-text-grey sub-menu"
          >
            <a
              href="/login"
              onClick={[Function]}
            >
              Login
            </a>
              · 
            <a
              href="../"
            >
              Forgot Password
            </a>
              · 
            <a
              href="../"
            >
              Need Help?
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>
</section>
`;
